src/Database/LSMTree/Internal/RefCount.hs

```diff
@@ -0,0 +1,115 @@
+
+{-# SPECIALISE mkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (Maybe (RefCounter IO)) #-}
+-- | Make a reference counter with initial value @n@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+--
+-- Returns 'Nothing' if @n < 1@, and 'Just' otherwise.
+mkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (Maybe (RefCounter m))
+mkRefCounterN !n finaliser
+  | n < RefCount 1 = pure Nothing
+  | otherwise = do
+    -- evaluate the finaliser a little bit before we store it
+    let !() = case finaliser of
+                Nothing       -> ()
+                Just !_action -> ()
+    countVar <- newMutVar n
+    finaliserVar <- newMutVar finaliser
+    pure $! Just $! RefCounter{countVar, finaliserVar}
+
+{-# SPECIALISE mkRefCounter1 :: Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Make a reference counter with initial value @1@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+mkRefCounter1 :: PrimMonad m => Maybe (m ()) -> m (RefCounter m)
+mkRefCounter1 finaliser = fromJust <$> mkRefCounterN (RefCount 1) finaliser
+
+data RefCountError = RefCountAlreadyZero
+  deriving stock (Show, Eq)
+  deriving anyclass Exception
+
+{-# SPECIALISE addReference :: RefCounter IO -> IO () #-}
+-- | Increase the reference counter by one.
+--
+-- If the reference counter was already at @0@, a 'RefCountError' error is
+-- thrown.
+addReference :: (PrimMonad m, MonadThrow m) => RefCounter m -> m ()
+addReference RefCounter{countVar} = do
+    success <- atomicModifyMutVar' countVar $ \c ->
+      if c == RefCount 0 then (c, False)
+                         else (incRefCount c, True)
+    unless success $ throwIO RefCountAlreadyZero
+
+{-# SPECIALISE readRefCount :: RefCounter IO -> IO RefCount #-}
+readRefCount :: PrimMonad m => RefCounter m -> m RefCount
+readRefCount RefCounter{countVar} = readMutVar countVar
+
+{-# SPECIALISE removeReference :: RefCounter IO -> IO () #-}
+-- | Decrease the reference counter by one.
+--
+-- When the reference counter reaches zero, the finaliser is run promptly.
+--
+-- If the reference counter was already at @0@, then this function is
+-- idempotent. The finaliser will not run again.
+removeReference :: PrimMonad m => RefCounter m -> m ()
+removeReference rc@RefCounter{countVar} = do
+    may <- atomicModifyMutVar' countVar $ \c ->
+      if c == RefCount 0 then (c, Nothing)
+                         else let c' = decRefCount c in (c', Just (c == RefCount 1))
+    case may of
+      Nothing -> pure ()
+      Just b  -> when b $ finalise rc
+
+{-# SPECIALISE replaceFinaliser :: RefCounter IO -> Maybe (IO ()) -> IO () #-}
+replaceFinaliser :: PrimMonad m => RefCounter m -> Maybe (m ()) -> m ()
+replaceFinaliser RefCounter{finaliserVar} finaliser =
+    writeMutVar finaliserVar finaliser
+
+{-# SPECIALISE finalise :: RefCounter IO -> IO () #-}
+finalise :: PrimMonad m => RefCounter m -> m ()
+finalise RefCounter{finaliserVar} = readMutVar finaliserVar >>= sequence_
```

Comment:
How about reading and resetting the finaliser to `Nothing` before running it? That way we would allow the thing the finaliser was finalising to become garbage, even if the recount itself is not yet garbage (because some other object has it, e.g. like a weak pointer style blob reference to the refcount of a run).

Reply:
No longer applicable, since the finaliser is no longer mutable